client_side/.infrastructure/verify_task.py

```diff
@@ -0,0 +1,188 @@
+
+        def __enter__(self):
+            self.savedPath = os.getcwd()
+            os.chdir(self.newPath)
+
+        def __exit__(self, etype, value, traceback):
+            os.chdir(self.savedPath)
+
+    # the true task code
+    task_code = sys.argv[1]
+    command_dir = sys.argv[2]
+    # the current command
+    command = ' '.join(sys.argv[3:])
+
+    try:
+
+        # Always:
+        # - Get the current state of the file system and compare it to the expected
+        #   file system for the current task.  (For a select task, this ensures that
+        #   it was not changed).
+        # If it is a "select" task and the file system was not modified, also:
+        # - Re-execute the user command and capture the `stdout`.
+        # - Check that the captured `stdout` of the user command matches the
+        #   corresponding expected output.
+
+        devnull = open(os.devnull, 'wb')
+
+        with open(USER_FS_FILE, 'w') as user_out:
+            with cd(FS_DIR):
+                filesystem = subprocess.call('find .', shell=True, stderr=devnull, stdout=user_out)
+
+        normalize_and_copy_output(USER_FS_FILE, ACTUAL_FILE)
+
+        # Verify checks whether or not the file system state is as expected.
+        fs_good = verify(ACTUAL_FILE, task_code, True)
+
+        if not fs_good:
+            if task_code in FILESYSTEM_TASKS:
+                sys.exit(FILE_SYSTEM_FAILURE)
+            else:
+                sys.exit(FILE_SYSTEM_MODIFIED)
+        else:
+            if task_code in FILESYSTEM_TASKS:
+                sys.exit(VERIFICATION_SUCCESS)
+            else:
+                with open(USER_STDOUT_FILE, 'w') as user_out:
+                    with open(USER_STDERR, 'w') as user_err:
+                        # shell=True is needed here for several reasons:
+                        # 1. It preserves quoting for user commands given that
+                        # the command argument passed to the call is a string.
+                        # (see https://docs.python.org/3/library/subprocess.html#popen-constructor starting at "On POSIX with shell=True")
+                        # 2. It reduces the need to replace shell piplines.
+                        # (see https://docs.python.org/3/library/subprocess.html#replacing-shell-pipeline)
+                        with cd(command_dir):
+                            stdout = subprocess.call(command, shell=True, stderr=user_err, stdout=user_out)
+
+                normalize_and_copy_output(USER_STDOUT_FILE, ACTUAL_FILE)
+
+                if verify(ACTUAL_FILE, task_code, False):
+                    sys.exit(VERIFICATION_SUCCESS)
+                else:
+                    sys.exit(STANDARD_OUT_FAILURE)
+        print("This can't happen")
+        sys.exit(UNEXPECTED_FAILURE)
+    except (OSError, subprocess.CalledProcessError) as e:
+        print(e)
+        sys.exit(SUBPROCESS_FAILURE)
+
+def normalize_and_copy_output(out_file, norm_file):
```

Comment:
I find the formal parameter `out_file` very confusing.  It is an input to this routine.  If anything were named `out_file`, it should be `norm_file`.  I see why you name it this way -- because clients of this routine will pass a file that is the output of some other command.  It's still confusing here.  You should write the specification of a routine from that routine's point of view, not with respect to some particular client.